src/visip/dev/dtype_new.py

```diff
@@ -0,0 +1,582 @@
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class Int(metaclass=Singleton):
+    pass
+
+
+class Float(metaclass=Singleton):
+    pass
+
+
+class Bool(metaclass=Singleton):
+    pass
+
+
+class Str(metaclass=Singleton):
+    pass
+
+
+class Class:
+    def __init__(self, module, name, origin_type):
+        self.module = module
+        self.name = name
+        self.origin_type = origin_type
+
+class Enum:
+    def __init__(self, module, name, origin_type):
+        self.module = module
+        self.name = name
+        self.origin_type = origin_type
+
+
+class List:
+    def __init__(self, arg):
+        self.arg = arg
+
+
+class Dict:
+    def __init__(self, key, value):
+        self.key = key
+        self.value = value
+
+
+class Tuple:
+    def __init__(self, args):
+        self.args = args
+
+
+class Union:
+    def __init__(self, args):
+        self.args = []
+
+        # expand nested unions
+        for a in args:
+            b = a
+            const = False
+            if isinstance(b, Const):
+                const = True
+                while isinstance(b, Const):
+                    b = b.arg
+            if isinstance(b, Union):
+                if const:
+                    ext = [Const(a) for a in b.args]
+                else:
+                    ext = a.args
+                self.args.extend(ext)
+            else:
+                self.args.append(a)
+
+        # remove duplicates
+        new_args = []
+        for a in self.args:
+            dup = False
+            for na in new_args:
+                if is_equaltype(na, a):
+                    dup = True
+                    break
+            if not dup:
+                new_args.append(a)
+        self.args = new_args
+
+        # check TypeVar count
+        tv_count = 0
+        tv_ind = -1
+        for i, a in enumerate(self.args):
+            if extract_type_var(a):
+                tv_count += 1
+                tv_ind = i
+                if tv_count > 1:
+                    raise TypeError("More than one argument with TypeVars in union.")
+
+        # place TypeVar argument to the end
+        if 0 <= tv_ind < len(self.args) - 1:
+            a = self.args.pop(tv_ind)
+            self.args.append(a)
+
+
+class Const:
+    def __init__(self, arg):
+        if isinstance(arg, Const):
+            self.arg = arg.arg
+        else:
+            self.arg = arg
+        assert not isinstance(self.arg, Const)
+
+
+class TypeVar:
+    def __init__(self, name):
+        self.name = name
+
+
+class NewType:
+    def __init__(self, name, supertype):
+        self.name = name
+        self.supertype = supertype
+
+
+class Any(metaclass=Singleton):
+    pass
+
+
+class NoneType(metaclass=Singleton):
+    pass
+
+
+def from_typing(type):
+    t, _, _ = from_typing_map(type, {}, {})
+    return t
+
+
+def from_typing_map(type, var_map, new_map):
```

Comment:
We should use conversion from typing only in action_def, but force usage of dtype_new in the VISIP code.

Reply:
Odlozit po merge, predelani vsech anotaci na dtype.